### PR TITLE
Add files via upload

### PR DESCRIPTION
**Pull request by zahcaryanstey**
Comments added and a working visualize prediction function

### CODE REVIEW

File: Aug 4 2021/Ch1ResnetV3.py [hunk 1]
```diff
@@ -1,245 +1,225 @@
...
+    padding: controls the amount of padding applied to the input. It can be either a string {'valid','same'} or a tuple of ints giving the amount of implicit padding applied on both sides 
+    in_channels = number of channels in the input image 
+    out_channels = number of channels produced by the convolution
+    kernel_size = Size of the convolving kernel 
+    stride = Stride of the convolution 
+    padding = padding added to both sides of the input. Default 0 
+    bias = if True, adds a learnable bias to the output. Default: True"""
+    model.fc = nn.Linear(in_features=512, out_features=1) # we alse had the change the fully connected layer so that it takes a (256 + 256 = 512) input and output just 1 feature because our problem is a regression problem
+    model.to(device) # send the model to the device defined above either GPU or CPU
     visualize_prediction(model.to(device),validation_loader)
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:122]: You don't need to sent it to(device) again here

File: Aug 4 2021/Ch1ResnetV3.py [hunk 2]
```diff
@@ -1,245 +1,225 @@
...
-    train_loader = DataLoader(train_set, batch_size=config.batch_size, shuffle=True, num_workers=config.num_workers)
-    validation_loader = DataLoader(validation_set, batch_size=config.batch_size, shuffle=False, num_workers=config.num_workers)
-
-    # Make the model
-    model = torchvision.models.resnet18(pretrained=True)
-    model.conv1 = nn.Conv2d(1, 64, (7, 2), padding=3, bias=False)
-    model.fc = nn.Linear(in_features=512, out_features=1)
-    model.to(device)
+"""Variables"""
+loss_list = [] # Empty list used to store loss during training list
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:46]: I would move the definition of loss_list to the begining of each training epoch, aka line 136. You want to define an empty list at the beginning of each epoch, append the loss to the list each step, and then average at the end of each epoch. E.g., if you train for 100 epochs, you should define loss_list 100 times and log it to wandb only 100 times.

File: Aug 4 2021/Ch1ResnetV3.py [hunk 3]
```diff
@@ -1,245 +1,225 @@
...
     optimizer.step()
-    return loss
-def train_log(loss, example_ct, epoch):
-    train_loss = float(loss)
 
+    return loss # return the loss
+def train_log(loss, example_ct, epoch): # define what we want to log with weights and biases
+    train_loss = float(loss) # convert the loss from a torch float tensor to a float
+    wandb.log({'train_per_sample_loss': train_loss}) # log the train per sample loss
+    wandb.log({'train_average_loss': np.mean(loss_list)}) # calculate and log the average loss
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:168]: see above comment, I don't think you want to log the average loss every timestep, this should only be called after the final iteration of each epoch, aka line 145 after the training for loop is done. You did it correctly for the Validation_average_loss (i.e., outside the for loop)

File: Aug 4 2021/Ch1ResnetV3.py [hunk 4]
```diff
@@ -1,245 +1,225 @@
...
-        #     plt.show()
-        #     plt.title('predicted:{}'.format(predicted[j]))
-        #     if images_so_far == num_images:
-        #         # model.train(model=was_training)
-        #         return
-
-    # model.train(mode=was_training)
 
 # Saving the model
 save_path = './mlp.pth'
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:221]: are you sure this is where you want to save it? I usually have a folder in my current directory with all my model names, such as:

save_path = "checkpoints/run_name/model_weights.pth"

and then you can save images or visualizations under "checkpoints/run_nume/visualizations/" or something. That way everything related to the training run called "run_name" is in the same folder, 'checkpoints/run_name/......."

File: Aug 4 2021/Ch1ResnetV3.py [hunk 5]
```diff
@@ -1,245 +1,225 @@
...
+"""
+def model_pipeline(hyperparameters): #function to tell weights and biases what to track
+
+    with wandb.init(project='test', config=hyperparameters): # tell weights and biases to get started
+        hyperparameter = wandb.config # access all the hyperparameters through the config dictionary and log them with weights and biases
+        model, train_loader, validation_loader, criterion, optimizer=make(hyperparameter)  # make the model, training data, validation data, and optimization and use them to train the model
+        wandb.watch(model, criterion, log='all', log_freq=10) # telling weights and biases what to watch
+        for epoch in range(hyperparameter.num_epochs):  # for epoch in range of the number of epochs specified above in the hyperparameter dictionary
+            train(model, train_loader, criterion, optimizer, hyperparameter) # train the model, function inputs = resnet model, training data, loss function (criterion), optimizer, and hyperparameter dictionary
+            validation(model, validation_loader, criterion) # validate the model, function inputs = resent model, validation data, and loss function (criterion)
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:77]: nice, this is a great structure for training over num_epochs

File: Aug 4 2021/Ch1ResnetV3.py (hunk 4 above)
Comment (MKowal2) [reply Aug 4 2021/Ch1ResnetV3.py:221]: Also, just a heads up, I am pretty sure "./" will save the file in your base directory, like if you type "cd ../" over and over and over you end up at this directory.

File: Aug 4 2021/Ch1ResnetV3.py [hunk 6]
```diff
@@ -1,245 +1,225 @@
...
+
+def visualize_prediction(model,validation_loader,num_images=10): # Function to visualize predictions made with the model. This function takes an input our model, validation data and the number of images you want to display
+      model.eval() # Evaluate the model
+      images_so_far = 0 # images that have been shown so far
+      with torch.no_grad(): # Using torch
+         for i,(data,targets) in enumerate(validation_loader): # for image in data and targets located in the validation loader
+             data = data.to(device) # send the images to the device
+             targets = targets.to(device) # send the labels to the device
+             outputs = model(data) # run the images through the model
+             _,predicted = torch.max(outputs,1) # returns the index of the highest prediction
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:206]: Do you want to return the index of the highest prediction here? This code looks like classification code, not regression. You are only predicting a float value. So I think this should be removed and you will be left with a single value for each image, which is what you want. (i.e., outputs.shape = batch x 1

File: Aug 4 2021/Ch1ResnetV3.py [hunk 7]
```diff
@@ -1,245 +1,225 @@
...
+             images_so_far +=1 # add one to the number of images used so far
+             image = data[i] # image = image tensor indexed at i
+             plt.imshow(image.cpu().squeeze(), cmap='gray') # plot the image
+             ground_truth = targets[i]
+             model_predicted = _[i]
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:211]: not sure what _ is, but I would clean this up and just use the variable "outputs" from your model, no need to deal with torch.max()

File: Aug 4 2021/Ch1ResnetV3.py [hunk 8]
```diff
@@ -1,245 +1,225 @@
...
+             title = ('ground truth',ground_truth,'model predicted',model_predicted) # set the title
+             plt.title(title) # add the title to the image
+             plt.show() # show the image
+             if images_so_far == num_images: # keep going over images until num_images = images_so_far
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:215]: can replace this logic with:

if i == num_images:
     return
     
     
no need to use the images_so_far tracker since the 'enumerate' iterator already does that

File: Aug 4 2021/Ch1ResnetV3.py [hunk 9]
```diff
@@ -1,245 +1,225 @@
...
-        #     if images_so_far == num_images:
-        #         # model.train(model=was_training)
-        #         return
-
-    # model.train(mode=was_training)
 
 # Saving the model
 save_path = './mlp.pth'
-torch.save(model_pipeline(config).state_dict(), save_path)
+torch.save(model_pipeline(hyperparameter).state_dict(), save_path)
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:222]: I think this is fine, but you might want to separate these out into a couple lines if you see a bug. But I see that model_pipeline returns the trained model and then u grab the state_dict, so this looks good to me

File: Aug 4 2021/Ch1ResnetV3.py [hunk 10]
```diff
@@ -1,245 +1,225 @@
...
+             outputs = model(data) # run the images through the model
+             _,predicted = torch.max(outputs,1) # returns the index of the highest prediction
+             images_so_far +=1 # add one to the number of images used so far
+             image = data[i] # image = image tensor indexed at i
+             plt.imshow(image.cpu().squeeze(), cmap='gray') # plot the image
+             ground_truth = targets[i]
+             model_predicted = _[i]
+             title = ('ground truth',ground_truth,'model predicted',model_predicted) # set the title
+             plt.title(title) # add the title to the image
+             plt.show() # show the image
```
Comment (MKowal2) [inline Aug 4 2021/Ch1ResnetV3.py:214]: you can change this to plt.save_fig and just save them and look at them after / during our meetings. Much easier than needing to run it each time and then they disapear
